apps/remote_control/lib/lexical/remote_control/code_mod/ast/aliases.ex

```diff
@@ -0,0 +1,197 @@
+defmodule Lexical.RemoteControl.CodeMod.Ast.Aliases do
+  defmodule Alias do
+    defstruct [:source, :dest, :ref]
+
+    def new(source, dest, ref) do
+      %__MODULE__{source: source, dest: dest, ref: ref}
+    end
+  end
+
+  defmodule Scope do
+    defstruct [:end_position, :ref]
+
+    def global do
+      %__MODULE__{ref: make_ref(), end_position: {:infinity, :infinity}}
+    end
+
+    def new(end_position) do
+      %__MODULE__{ref: make_ref(), end_position: end_position}
+    end
+
+    def ended?(%__MODULE__{end_position: {:infinity, :infinity}}, _) do
+      false
+    end
+
+    def ended?(%__MODULE__{} = scope, {line, column}) do
+      {end_line, end_column} = scope.end_position
+
+      if line == end_line do
+        column >= end_column
+      else
+        line > end_line
+      end
+    end
+  end
+
+  defmodule State do
+    defstruct aliases: %{}, scopes: [], last_elem: nil
+
+    def new do
+      %__MODULE__{scopes: [Scope.global()]}
+    end
+
+    def update(%__MODULE__{} = state, elem) do
+      state
+      |> maybe_pop_scope(elem)
+      |> do_update(elem)
+      |> Map.put(:last_elem, elem)
+    end
+
+    defp do_update(%__MODULE__{} = state, {:defmodule, metadata, _body}) do
+      push_scope(state, metadata)
+    end
+
+    defp do_update(%__MODULE__{} = state, {:def, metadata, _body}) do
+      push_scope(state, metadata)
+    end
+
+    defp do_update(%__MODULE__{} = state, {:__aliases__, _meta, module_alias}) do
+      case state.last_elem do
+        {:defmodule, _, _} ->
+          module_alias =
+            case Map.fetch(state.aliases, :__MODULE__) do
+              {:ok, alias_struct} ->
+                Module.split(alias_struct.source) ++ module_alias
+
+              _ ->
+                module_alias
+            end
+
+          put_alias(state, module_alias, :__MODULE__)
+
+        _ ->
+          state
+      end
+    end
+
+    # A simple alias: alias Foo.Bar
+    defp do_update(
+           %__MODULE__{} = state,
+           {:alias, _metadata, [{:__aliases__, _, aliased_module}]}
+         ) do
+      source = ensure_alias(aliased_module)
+      dest = aliased_module |> List.last() |> ensure_alias()
+      put_alias(state, source, dest)
+    end
+
+    # An alias with a specified name: alias Foo.Bar, as: FooBar
+
+    defp do_update(
+           %__MODULE__{} = state,
+           {:alias, _metadata,
+            [{:__aliases__, _, aliased_module}, [as: {:__aliases__, _, dest_alias}]]}
+         ) do
+      source = ensure_alias(aliased_module)
```

Comment:
Since we have already performed the ensure action in put_alias, I think there are many places where we don't need to ensure anymore.